pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,383 @@
+		ownerNode.dependents = append(ownerNode.dependents, n)
+	}
+	return ownerExists
+}
+
+// TODO: finish this function.
+func equalReferences(a []metatypes.OwnerReference, b []metatypes.OwnerReference) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	// sort and compare
+	return true
+}
+
+// Dequeue from eventQueue, updating graph, populating dirty_queue.
+func (p *Propagator) processEvent() {
+	key, quit := p.eventQueue.Get()
+	if quit {
+		return
+	}
+	defer p.eventQueue.Done(key)
+	event, ok := key.(Event)
+	if !ok {
+		glog.Errorf("expect an Event, got %v", key)
+		return
+	}
+	obj := event.Obj
+	accessor, err := meta.Accessor(obj)
+	typeAccessor, err := meta.TypeAccessor(obj)
+	if err != nil {
+		glog.Errorf("cannot access obj: %v", err)
+		return
+	}
+	glog.V(6).Infof("Propagator process object: %s/%s, namespace %s, name %s, event type %s", typeAccessor.GetAPIVersion(), typeAccessor.GetKind(), accessor.GetNamespace(), accessor.GetName(), event.Type)
+	switch event.Type {
+	case Add:
+		// Check if the node already exsits
+		_, ok := p.graphLookup[accessor.GetUID()]
+		if ok {
+			// We can optionally check if the node in the graph is in sync with that in the Store.
+			return
+		}
+		newNode := &node{
+			identity: SelfReference{
+				OwnerReference: metatypes.OwnerReference{
+					APIVersion: typeAccessor.GetAPIVersion(),
+					Kind:       typeAccessor.GetKind(),
+					UID:        accessor.GetUID(),
+					Name:       accessor.GetName(),
+				},
+				Namespace: accessor.GetNamespace(),
+			},
+			owners: accessor.GetOwnerReferences(),
+		}
+		p.addToGraphLookup(newNode)
+		ownerExists := p.updateOwners(newNode)
+		// Push the object to the dirty queue if none of its owners exists in the Graph.
+		if !ownerExists && len(newNode.owners) != 0 {
+			p.gc.dirtyQueue.Add(newNode)
+		}
+
+	case Update:
+		node, ok := p.graphLookup[accessor.GetUID()]
+		if !ok {
+			glog.Errorf("received an update for %v, but cannot find the node in the graph", accessor.GetUID())
+		}
+		// TODO: finalizer: Check if ObjectMeta.DeletionTimestamp is updated from nil to non-nil
+		// We only need to add/remove owner refs for now
+		if !equalReferences(node.owners, accessor.GetOwnerReferences()) {
+			ownerExists := p.updateOwners(node)
+			if !ownerExists && len(node.owners) != 0 {
+				p.gc.dirtyQueue.Add(node)
+			}
+		}
+	case Delete:
+		node, ok := p.graphLookup[accessor.GetUID()]
+		if !ok {
+			glog.V(6).Infof("%v doesn't exist in the graph, this shouldn't happen", accessor.GetUID())
+		}
+		p.removeFromGraphLookup(node)
+		for _, dep := range node.dependents {
+			p.gc.dirtyQueue.Add(dep)
+		}
+	}
+}
+
+type GarbageCollector struct {
+	restMapper meta.RESTMapper
+	clientPool dynamic.ClientPool
+	dirtyQueue *workqueue.Type
+	monitors   []Monitor
+	propagator *Propagator
+}
+
+func monitorFor(p *Propagator, clientPool dynamic.ClientPool, resource unversioned.GroupVersionResource) (Monitor, error) {
+	// TODO: consider store in one storage.
```

Comment:
I'm worried that we're going to be keeping the contents of every object in the system in memory. Maybe you can construct the dynamic clients with a special codec which doesn't actually store the contents of the Unstructured objects, just the meta info?